internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +75,53 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
 
 // PlanChanged handles updates to the plan (server configuration),
 // stopping the previous checks and starting the new ones as required.
-func (m *CheckManager) PlanChanged(p *plan.Plan) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-
-	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
-		len(m.checks), len(p.Checks))
-
-	// First stop existing checks.
-	for _, check := range m.checks {
-		check.cancel()
-	}
-	// Wait for all context cancellations to propagate and allow
-	// each goroutine to cleanly exit.
-	m.wg.Wait()
-
-	// Set the size of the next wait group
-	m.wg.Add(len(p.Checks))
-
-	// Then configure and start new checks.
-	checks := make(map[string]*checkData, len(p.Checks))
-	for name, config := range p.Checks {
-		ctx, cancel := context.WithCancel(context.Background())
-		check := &checkData{
-			config:  config,
-			checker: newChecker(config, p),
-			ctx:     ctx,
-			cancel:  cancel,
-			action:  m.callFailureHandlers,
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
+	// TODO: should figure out how to not stop/restart checks that haven't changed
+	//       so that unnecessary changes/tasks aren't created
+
+	// First cancel tasks of currently-running checks.
+	m.checksLock.Lock()
+	for name, data := range m.checks {
+		if data.cancel != nil {
+			data.cancel()
 		}
```

Comment:
I assume this mechanism is in place, as apposed to ```change.Abort()```, which would require us to use the current change or change ID to abort, so we have a single way to cancel whichever change is currently in flight.

Would it not be possible for a transition between prepare and recover to sometimes take place even with the context cancelled (i.e. tomb in dying state) ? I guess worse case on arrival, the check would have been deleted from the map, and would then exit in any case.

Reply:
> Would it not be possible for a transition between prepare and recover to sometimes take place even with the context cancelled (i.e. tomb in dying state) ? I guess worse case on arrival, the check would have been deleted from the map, and would then exit in any case, so actually it looks fine.

Oh, this is a good point. It's actually worse than that, because the check isn't necessarily in the map yet. I'd love to brainstorm with you (maybe tomorrow morning my time) a better way to approach this.

Reply:
I've reworked this significantly so all the data is stored in change/task data, so there's only one copy (with the exception of the "health" cache for the `/v1/health` endpoint).